### ISSUE TITLE
Conditionally typed parameter and unions not enforced

### ISSUE THREAD

**Issue by jeffijoe** [microsoft/TypeScript#39786]
**TypeScript Version:**  3.9.2

**Search Terms:** typescript pass callback undefined conditional type

**Expected behavior:** passing a callback that specifies `T` as the parameter type to a target that specifies `T | undefined` after evaluating the conditional type should emit an error.

**Actual behavior:** passing a callback that specifies `T` as the parameter type to a target that specifies `T | undefined` after evaluating the conditional type does not emit an error, resulting in potentially attempting to access properties of `undefined`. By wrapping it in `(v) => callback(v)` however, the behavior is correct.

**Related Issues:** Some that might have the opposite problem? Not sure though. https://github.com/microsoft/TypeScript/issues/26013

**Code**
```ts
interface Params<T, B extends boolean> {
    shouldUseArray: B
    callback(value: B extends true ? T[] : T | undefined): void
}

declare function fn<T, B extends boolean>(value: T, params: Params<T, B>): void

declare function myCallback(num: number): void

const input: number = 123

fn(input, {
    shouldUseArray: false,
    // good, the undefined check is enforced
    callback: v => myCallback(v)
                          //  ^ possible undefined, so this works
})

fn(input, {
    shouldUseArray: false,
    // bad, the undefined check is not enforced
    callback: myCallback
})

```

<details><summary><b>Output</b></summary>

```ts
"use strict";
const input = 123;
fn(input, {
    shouldUseArray: false,
    // good, the undefined check is enforced
    callback: v => myCallback(v)
});
fn(input, {
    shouldUseArray: false,
    // bad, the undefined check is not enforced
    callback: myCallback
});

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=23&pc=1#code/JYOwLgpgTgZghgYwgAgApynAtgZwDwAqANMgELIQAekIAJjsgEYD2zANhHCAHzIDeAKGTDkOABbMArm1oBVHBACCUTAE8AXGSEiEcNm0aIA1gAoAbnskRN5KjXrIwUK8gD8yAgG0Ausk0FkAB9kSToIGFAIWgBKTTNmYFoBAF8BAVoIBDYMFBhQhDBgZhBkGBBCEltqCDoGFnZOHnNLaw8SAAcMbBxNdExcCrJuWOR4xLSMrJzS-MLi5CxVAGE9A2MTEEksTU2sRmgRsaSBBGKcMGRQdskwHa39qGQAXmQARgAmAGY0spMrm5IghEogk0jkCmUak08DYCiI2mEAHpEcgAOasWgkMBiFChDIREBRZAIHEIIyXBg1GDMKBIJLA3T6QxkuLPXiLFZM9ZmaIpXkCX7-MCAhEgqQyeRKFRwDSlPRw0XIphwTGOHEhMIEokkzLk4AMEDMC5Uml00WMtYshbLVbMox8gRAA)

**Comment by RyanCavanaugh** [comment]
Functions declared with method syntax are checked bivariantly. You should write it as
```ts
    callback: (value: B extends true ? T[] : (T | undefined)) => void
```
if you want contravariant parameter checking, which you do.

**Comment by jeffijoe** [comment]
Interesting, I didn't know that! I always thought the 2 were equivalent.

What's the rationale behind that? Is it documented somewhere?

**Comment by RyanCavanaugh** [comment]
See comments in https://github.com/microsoft/TypeScript/pull/18654

We need to add this to the reference pages in the new TS Handbook. Unfortunately this is the kind of thing where no one goes digging through the docs to discover it until *after* they've already noticed, at which point it's already sort of too late.
